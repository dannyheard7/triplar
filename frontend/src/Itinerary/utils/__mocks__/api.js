const faker = require('faker');

export default {
    curryAPIFunc (apiFunc, id) {return (object) => apiFunc(id, object)},

    searchCities(value) {
        return Promise.resolve({status: 200, data: {data: {cities: [faker.address.city(),  faker.address.city()]}}});
    },

    addLocationToTrip(tripId, value) {
        return Promise.resolve({status: 200, data: value});
    }
}