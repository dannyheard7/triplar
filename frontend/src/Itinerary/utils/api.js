import axios from "axios";
import Q from "q";
import {backendHost} from 'App/utils/constants';

export default {
    curryAPIFunc (apiFunc, id) {return (object) => apiFunc(id, object)},

    addLocationToTrip(tripId, destination) {
        const mutateTripLocationQuery = ` 
            mutation AddTripLocation($input: TripLocationMutationInput!){
              result: addTripLocation(input: $input) {
                tripLocation {
                    id
                    startDate
                    endDate
                    city {
                        name
                        location
                    }
                    trip {
                        id
                    }
                }
                errors {
                  field
                  messages
                }
              }
            }`;


        let variables = {input: {trip: tripId, ...destination}};

        return Q.when(axios.post(`${backendHost}/api/graphql`, {query: mutateTripLocationQuery, variables: variables}));
    },

    getLocationDayItinerary(tripLocationId, date) {
        const getLocationDayItinerary = ` 
           query { 
            locationDayItinerary(date: "${date}", locationId: ${tripLocationId}) {
                place {
                    id
                    name
                    imageUrl
                }
            }
           }`;

        return Q.when(axios.post(`${backendHost}/api/graphql`, {query: getLocationDayItinerary}));
    },

    addItemToTripLocation(tripLocationId, placeId, date, order) {
        const addTripLocationItemMutation = ` 
            mutation AddTripLocationItem($input: AddTripLocationItemMutationInput!){
              addTripLocationItem(input: $input) {
                tripLocationItem {
                  order
                }
                errors {
                  field
                  messages
                }
              }
            }`;

        let variables = {input: {location: tripLocationId, apiPlace: placeId, order: order, startDate: date, endDate: date}};

        return Q.when(axios.post(`${backendHost}/api/graphql`, {query: addTripLocationItemMutation, variables: variables}));
    },

    searchCities(query) {
        const searchCities = ` 
           query { 
            cities(name: "${query}") {
                name
                country {
                    name
                }
            }
           }`;

        return Q.when(axios.post(`${backendHost}/api/graphql`, {query: searchCities}));
    },
}