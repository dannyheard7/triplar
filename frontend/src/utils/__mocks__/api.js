const faker = require('faker');

export default {

    getTrips() {
        return new Promise((resolve, reject) => {
            let trips = [faker.random.word(), faker.random.word()];
            let response = {status: 200, data: trips};
            resolve(response);
        })
    },

    createTrip(trip) {
        return new Promise((resolve, reject) => {
            let response = {status: 201, data: trip};
            resolve(response)
        })
    },

    getTrip(id)  {
         const trip = {
            'id': id, 'name': faker.random.word(), 'created_by': faker.internet.email()
        };
        return new Promise((resolve, reject) => {
            let response = {status: 200, data: trip};
            resolve(response);
        });
    }
}