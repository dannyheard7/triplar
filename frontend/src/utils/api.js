import axios from "axios";
import Q from "q";

export default {
    getTrips() {
        return Q.when(axios.get("/trips/"));
    },

    createTrip(trip) {
        return Q.when(axios.post("/trips/", trip));
    },

    getTrip(tripId) {
        return Q.when(axios.get("/trips/" + tripId));
    },

    patchTrip(trip) {
        return Q.when(axios.patch("/trips/" + trip.id, trip));
    },

    deleteTrip(tripId) {
        return Q.when(axios.delete("/trips/" + tripId));
    }
}