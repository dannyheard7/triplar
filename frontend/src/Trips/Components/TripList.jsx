import React from "react";
import {Link} from "react-router-dom";
import Moment from "moment";

import '../styles/trips.css';
import {getTrips} from "../utils/actions";
import { connect } from "react-redux";

export class TripList extends React.Component {
    componentDidMount() {
        this.props.dispatch(getTrips())
    }

    render() {
        return (
            <div>
                <h1>Your Trips</h1>
                <div className="trip-list">
                    {this.props.trips.map(this.createListItem)}
                </div>
            </div>
        )
    }

    createListItem(trip) {
        return (
            <div className="card trip-card" key={trip.id}>
                <h3 className="card-title"><Link to={"/trips/" + trip.id}>{trip.name}</Link></h3>
                <p className="card-text">
                    <span>{trip.locations && trip.locations.map((location, index) => (index ? ', ': '') + location.city.name)}</span><br />
                    {Moment(trip.startDate).format('Do MMMM') } - {Moment(trip.endDate).format('Do MMMM') }
                </p>
            </div>
        );
    }
}

TripList.defaultProps = {
    trips: []
};

const mapStateToProps = state => ({
  trips: state.trips.trips,
});

export default connect(mapStateToProps)(TripList)