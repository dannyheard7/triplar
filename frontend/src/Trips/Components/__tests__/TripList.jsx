import React from "react";
import {shallow} from "enzyme";

import {TripList} from "../TripList";
import api from "../../utils/trips.api";
import ShallowRenderer from "react-test-renderer/shallow";

const renderer = new ShallowRenderer();
const faker = require('faker');


jest.mock('../../utils/trips.api.js');
describe('<TripList />', () => {
    describe('with no trips prop',() => {
        it('displays no <ListGroupItem /> component', () => {
            const container = shallow(<TripList dispatch={jest.fn()} />);
            expect(container.find(".trip-card").length).toEqual(0);
        });
    });

    describe('with trips prop',() => {
        const props = {
            trips: [],
            dispatch: jest.fn()
        }

        beforeEach(function() {
            props.trips = [{'id': faker.random.number(), 'name': faker.random.word(), 'locations': [{city: {name: faker.address.city()}}]}];
        });

        it('displays a <ListGroupItem /> component for a single trip', () => {
            const container = shallow(<TripList {...props} />);
            expect(container.find(".trip-card").length).toEqual(props.trips.length);
        });

        it('updates on new trip props', () => {
            const container = shallow(<TripList {...props} />);
            expect(container.find(".trip-card").length).toEqual(props.trips.length);

            let newTrips = [{'id': faker.random.number(), 'name': faker.random.word(), 'locations': [{city: {name: faker.address.city()}}]},
                {'id': faker.random.number(), 'name': faker.random.word(), 'locations': [{city: {name: faker.address.city()}}]}];
            container.setProps({trips: newTrips});
            expect(container.find(".trip-card").length).toEqual(newTrips.length);
        });
    });

    const props = {
        dispatch: jest.fn(),
        trips: []
    };

    beforeEach(() => {
        props.dispatch.mockReset();
    })

    test('renders a correctly', () => {
        const result = renderer.render(<TripList {...props} />);
        expect(result).toMatchSnapshot();
    });

    test('dispatches getTrips upon mount', () => {
        shallow(<TripList {...props} />);
        expect(props.dispatch).toBeCalled();
    });

    // This can be moved to the redux tests
    test.skip('update sets trips returned from server as state', async () => {
        let trips = [faker.random.word(), faker.random.word()];

        // Need to refactor into a better way of changes mocks for a single test
        api.getTrips = jest.fn().mockReturnValueOnce(new Promise((resolve, reject) => {
            let response = {status: 200, data: {data: {trips: trips}}};
            resolve(response);
        }));

        const container = shallow(<TripList />);
        await Promise.resolve();

        expect(container.state('trips')).toEqual(trips);
    });
});