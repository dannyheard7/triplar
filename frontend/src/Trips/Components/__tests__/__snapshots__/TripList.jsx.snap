// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TripList /> renders a correctly 1`] = `
<div>
  <h1>
    Your Trips
  </h1>
  <div
    className="tripList"
  />
</div>
`;
